### ISSUE TITLE
[BUG] [INTEG] Dockerhub ratelimiting

### ISSUE THREAD

**Issue by turan18** [awslabs/soci-snapshotter#385]
**Describe the bug**

Most integration tests run in their own container. Each container ( thus each test) usually ends up pulling our test image set from Dockerhub to build soci indices and ztocs.

Eg;
https://github.com/awslabs/soci-snapshotter/blob/75f860e054206960776f5a7a43337d238faddd82/integration/index_test.go#L103

https://github.com/awslabs/soci-snapshotter/blob/75f860e054206960776f5a7a43337d238faddd82/integration/create_test.go#L66

https://github.com/awslabs/soci-snapshotter/blob/6d83eef8cadf1c299050cd539b002fd19a0fe99f/integration/ztoc_test.go#L68

If we continue to add more tests there is a likely possibility of getting rate limited by Dockerhub.

**Expected behavior**

ideally, we would like each test container to have a copy of the image set without having to explicitly pull them. 

One possible way to do this would be to have each of the 4 test images pulled before hand on the users host (either using docker-cli or some containerd client). After pulling the images you can save each as a tarball using something `docker save`. You can then copy these tarballs over to the Dockerfile used to create the test containers like so `COPY img1.tar img2.tar img3.tar img4.tar soci_test_images/`. At this point each container can `docker load` those images from the `soci_test_images/` directly without having to pull them. 

#344 (In review) Introduces a `setup()` function that runs before integration tests, so this would be an ideal place to pull the image set and generate the tarballs.

The only caveat to this approach is that the containers themselves do not currently have docker installed in them. In order to load the image set in the container you will need a dind setup (Docker-in-Docker) as the containerd image import API does not work with docker image tarballs. We can get around this by either:

1. Using `ctr export` or the go client equivalent instead of `docker save` on the host
2. Installing docker within each container (dind)

**Configuration (please complete the following information):**
 - OS: 
 - Snapshotter Version:
 - Containerd Version: 

**Additional context**

**Comment by sparr** [comment]
Could we run a repository on the host (likely in another container) which the test containers could `ctr image pull` from instead of using docker or `ctr import`?

**Comment by sparr** [comment]
Would it be useful/reasonable to have an integration test flag that skips downloading the images if there's already a version of them on the host machine? This would speed up the tests for common usage, and reduce network usage, at the expense of making them less authoritative when run with that flag. That might deserve to be a separate feature request once this one lands.

**Comment by rdpsin** [comment]
Is this an issue? The `dockerhub` function name is misleading, because it's really pulling from ECR Public. 

https://github.com/awslabs/soci-snapshotter/blob/6a4ae7d10e9d5d7cc575c977c36a7a17420b5634/integration/testutil.go#L68


https://github.com/awslabs/soci-snapshotter/blob/6a4ae7d10e9d5d7cc575c977c36a7a17420b5634/integration/testutil.go#L186-L192

**Comment by turan18** [comment]
> Would it be useful/reasonable to have an integration test flag that skips downloading the images if there's already a version of them on the host machine? This would speed up the tests for common usage, and reduce network usage, at the expense of making them less authoritative when run with that flag. That might deserve to be a separate feature request once this one lands.

The issue is that the pulling happens within the test containers not within the host, so there needs to be a mechanism to have the test contianer have a copy of those images.

**Comment by sparr** [comment]
What I meant is that if we use a solution that copies images from the host to the containers, or where the containers pull from the host (instead of from docker), then we will probably still default to the host pulling the images once. That might not always be necessary, when desiring fast test iteration loops; the host could just provide to the containers whatever version of the images it has handy.

**Comment by turan18** [comment]
> Is this an issue? The `dockerhub` function name is misleading, because it's really pulling from ECR Public.
> 
> IIRC, we use ECR Public because of this exact issue. We were hitting Dockerhub rate limits, so we switched.
> 
> https://github.com/awslabs/soci-snapshotter/blob/6a4ae7d10e9d5d7cc575c977c36a7a17420b5634/integration/testutil.go#L68
> 
> https://github.com/awslabs/soci-snapshotter/blob/6a4ae7d10e9d5d7cc575c977c36a7a17420b5634/integration/testutil.go#L186-L192

Edited description to be more reflective of the ask

**Comment by turan18** [comment]
> Is this an issue? The `dockerhub` function name is misleading, because it's really pulling from ECR Public.
> 
> IIRC, we use ECR Public because of this exact issue. We were hitting Dockerhub rate limits, so we switched.
> 
> https://github.com/awslabs/soci-snapshotter/blob/6a4ae7d10e9d5d7cc575c977c36a7a17420b5634/integration/testutil.go#L68
> 
> https://github.com/awslabs/soci-snapshotter/blob/6a4ae7d10e9d5d7cc575c977c36a7a17420b5634/integration/testutil.go#L186-L192

Edited issue to be more reflective of the ask/issue

**Comment by sparr** [comment]
This issue is on hold until we discuss if/how to refactor the container-in-container pattern of the integration tests. Some possible outcomes of that process could make this issue moot.
